File: 2021/day_1/day_1.py
from operator import concat
import numpy as np
import pandas as pd

# part 1
d = np.loadtxt("2021/day_1/input.txt")
d1 = d[1:len(d)]
d1 = np.append(d1, 0)

ans1 = sum(d < d1)

# part 2
dd = pd.read_table("2021/day_1/input.txt", names=["depths"])

dd["sum_roll3"] = dd["depths"].rolling(3).sum()

dd["lag"] = dd["sum_roll3"].shift(1)
dd["increase"] = dd["sum_roll3"] > dd["lag"]

ans2 = dd["increase"].sum()


print(f'The answer to Part 1 is {ans1}.\nThe answer to Part 2 is {ans2}.')

File: 2021/day_1/day_1.R
library(tidyverse)
library(slider)


# Part 1 ------------------------------------------------------------------


d <- read_table("2021/day_1/input.txt", col_names = "depths") 

d %>% 
  mutate(lag = slide_dbl(depths, ~.x, .before = 1, 
                         .after = -1, .complete = T),
         increase = depths > lag) %>% 
  summarise(num_increases = sum(increase, na.rm = T))


# Part 2 ------------------------------------------------------------------


d %>% 
  mutate(sum_3 = slide_dbl(depths, sum, 
                           .before = 0, .after = 2, 
                           .complete = T),
         lag_sum_3 = slide_dbl(sum_3, ~.x, 
                               .before = 1, .after = -1, 
                               .complete = T),
         increase = sum_3 > lag_sum_3) %>% 
  summarise(num_increases = sum(increase, na.rm = T))


# shortest -------------------------------------------------------------------------

sum(lag(d$depths) < d$depths, na.rm = T)

d <- d %>% 
  mutate(sum_3 = slide_dbl(depths, sum, 
                           .before = 0, .after = 2, 
                           .complete = T))

sum(lag(d$sum_3) < d$sum_3, na.rm = T)


# Emil Hvitfeldt's approach -------------------------------------------------

input <- scan("2021/day_1/input.txt")

sum(diff(input, lag = 1) > 0)

sum(diff(input, lag = 3) > 0)

# you don't actually need to compare the whole sums of the windows. Two overlapping windows share all the intermediate values, it's just the first and last values that actually differ, so you can just compare those directly. So you just get the differences between lagged values